docs/merkletree/treetex/main.go

```diff
@@ -239,6 +259,99 @@ func toNodeKey(n storage.NodeID) string {
 	return nodeKey(d, i)
 }
 
+// decompose splits out a range into component subtrees.
+// TODO(al): Remove this and depend on actual range code.
+func decomposeRange(begin, end uint64) (uint64, uint64) {
+	// Special case, as the code below works only if begin != 0, or end < 2^63.
+	if begin == 0 {
+		return 0, end
+	}
+	xbegin := begin - 1
+	// Find where paths to leaves #begin-1 and #end diverge, and mask the upper
+	// bits away, as only the nodes strictly below this point are in the range.
+	d := bits.Len64(xbegin^end) - 1
+	mask := uint64(1)<<uint(d) - 1
+	// The left part of the compact range consists of all nodes strictly below
+	// and to the right from the path to leaf #begin-1, corresponding to zero
+	// bits in the masked part of begin-1. Likewise, the right part consists of
+	// nodes below and to the left from the path to leaf #end, corresponding to
+	// ones in the masked part of end.
+	return ^xbegin & mask, end & mask
+}
+
+// parseRanges parses and validates a string of comma-separates open-closed
+// ranges of the form L:R.
+// Returns the parsed ranges, or an error if there's a problem.
+func parseRanges(ranges string, treeSize int64) ([][2]int64, error) {
+	rangePairs := strings.Split(ranges, ",")
+	numRanges := len(rangePairs)
+	if num, max := numRanges, maxRanges; num > max {
+		return nil, fmt.Errorf("too many ranges %d, must be %d or fewer", num, max)
+	}
+	ret := make([][2]int64, 0, numRanges)
+	for _, rng := range rangePairs {
+		lr := strings.Split(rng, ":")
+		if len(lr) != 2 {
+			return nil, fmt.Errorf("specified range %q is invalid", rng)
+		}
+		var l, r int64
+		if _, err := fmt.Sscanf(rng, "%d:%d", &l, &r); err != nil {
+			return nil, fmt.Errorf("range %q is malformed: %s", rng, err)
+		}
+		switch {
+		case r > treeSize:
+			return nil, fmt.Errorf("range %q extends past end of tree (%d)", lr, treeSize)
+		case l < 0:
+			return nil, fmt.Errorf("range %q has negative beginning", rng)
+		case l > r:
+			return nil, fmt.Errorf("range elements in %q are out of order", rng)
+		}
+		ret = append(ret, [2]int64{l, r})
+	}
+	return ret, nil
+}
+
+// modifyRangeNodeInfo sets style info for nodes affected by ranges.
+// This includes leaves and perfect subtree roots.
+// TODO(al): Figure out what, if anything, to do to make this show ranges
+// which are inside the perfect meganodes.
+func modifyRangeNodeInfo() error {
+	rng, err := parseRanges(*ranges, *treeSize)
+	if err != nil {
+		return err
+	}
+	for ri, lr := range rng {
+		l, r := lr[0], lr[1]
+		// Set leaves:
+		for i := l; i < r; i++ {
+			modifyNodeInfo(nodeKey(0, i), func(n *nodeInfo) { n.rangeIndices = append(n.rangeIndices, ri) })
+		}
+
+		// Now perfect roots which comprise the range:
+		maskL, maskR := decomposeRange(uint64(l), uint64(r))
+		p := l
+		// Do left perfect subtree roots:
+		nBitsL := uint64(bits.Len64(maskL))
+		for i, bit := uint64(0), uint64(1); i < nBitsL; i, bit = i+1, bit<<1 {
```

Comment:
Does `i` have to be 64 bit? Could be `uint`. Same for `nBitsL`.

Reply:
done